src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -515,6 +524,64 @@ object ICustomAstVisitor.VisitArrayLiteral(ArrayLiteralAst arrayLiteralAst)
 
         object ICustomAstVisitor.VisitHashtable(HashtableAst hashtableAst)
         {
+            if (hashtableAst.KeyValuePairs.Count > 0)
+            {
+                var properties = new List<PSMemberNameAndType>();
+
+                foreach (var kv in hashtableAst.KeyValuePairs)
+                {
+                    string name = null;
+                    string typeName = null;
+                    if (kv.Item1 is StringConstantExpressionAst stringConstantExpressionAst)
+                    {
+                        name = stringConstantExpressionAst.Value;
+                    }
+                    else if (kv.Item1 is ConstantExpressionAst constantExpressionAst)
+                    {
+                        name = constantExpressionAst.Value.ToString();
+                    }
+                    else if (SafeExprEvaluator.TrySafeEval(kv.Item1, _context.ExecutionContext, out object nameValue))
+                    {
+                        name = nameValue.ToString();
+                    }
+                    if (name != null)
+                    {
+                        object value = null;
+                        if (kv.Item2 is PipelineAst pipelineAst
+                            && pipelineAst.PipelineElements.Count == 1
+                            && pipelineAst.PipelineElements[0] is CommandExpressionAst commandExpressionAst)
```

Comment:
We can use `pipelineAst.GetPureExpression()` here, which does the same thing and return null if it's not a pure expression. pipelineAst.GetPureExpression()
```c#
if (kv.Item2 is PipelineAst pipelineAst
    && pipelineAst.GetPureExpression() is ExpressionAst expression)
{
      switch (expression)
      {
          ....
```

Reply:
Didn't know about PureExpression. Excellent!

Reply:
That also covers the redirection case. Nice!